docs/compilers/CSharp/Runtime Async Design.md

```diff
@@ -0,0 +1,456 @@
+compiled with runtime async or not.
+
+## Supporting runtime apis
+
+We use the following runtime flag to drive whether feature can be used. This flag must be defined in the same assembly that defines `object`, and the assembly cannot reference any other assemblies. In terms of
+CoreFX, this means it must be defined in the `System.Runtime` reference assembly.
+
+TODO: Determine whether just the presence of this flag will cause the compiler to generate in runtime async mode.
+
+```cs
+namespace System.Runtime.CompilerServices;
+
+public static class RuntimeFeature
+{
+    public const string Async = nameof(Async);
+}
+```
+
+We use the following helper APIs to indicate suspension points to the runtime, in addition to the runtime async call syntax:
+
+```cs
+namespace System.Runtime.CompilerServices;
+
+// These methods are used to await things that cannot use runtime async signature form
+// TODO: Clarify which of these should be preferred? Should we always emit the `Unsafe` version when awaiting something that implements `ICriticalNotifyCompletion`?
+namespace System.Runtime.CompilerServices;
+
+public static class RuntimeHelpers
+{
+    [RuntimeAsyncMethod]
+    public static Task AwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : INotifyCompletion;
+    [RuntimeAsyncMethod]
+    public static Task UnsafeAwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : ICriticalNotifyCompletion;
+}
+```
+
+Additionally, we use the following helper attributes to indicate information to the runtime. If these attributes are not present in the reference assemblies, we will generate them; the runtime matches by full
+name, not by type identity, so we do not need to care about using the "canonical" versions.
+
+```cs
+namespace System.Runtime.CompilerServices;
+
+// Used to tell the runtime to generate the async state machinery for this method
+[AttributeUsage(AttributeTargets.Method)]
+public class RuntimeAsyncMethodAttribute() : Attribute();
+
+// Used to mark locals that should be hoisted to the generated async closure. Note that the runtime does not guarantee that all locals marked with this attribute will be hoisted; if it can prove that it
+// doesn't need to hoist a variable, it may avoid doing so.
+public class HoistedLocal();
+```
+
+For experimentation purposes, we recognize an attribute that can be used to force the compiler to generate the runtime async code, or to force the compiler to generate a full state machine. This attribute is not
+defined in the BCL, and exists as an escape hatch for experimentation. It may be removed when the feature ships in stable.
+
+```cs
+namespace System.Runtime.CompilerServices;
+
+[AttributeUsage(AttributeTargets.Method)]
+public class RuntimeAsyncMethodGenerationAttribute(bool runtimeAsync) : Attribute();
+```
+
+## Transformation strategy
```

Comment:
It would be good to describe the transformation strategy for `await`, which would explain the examples before.
From my understanding we have the following cases:
1. invocation of method returning the special types Task/Task<T>: emit as a call, but with special signature/convention
2. all other cases: use helper

Consider calling out that compiler-async methods require no changes, and can call and use the results of runtime-async methods.